# Quiz 4<br><small>Joshua Nasiatka (12.08.16)</small>
### Problem 1:
IMDB.com is the Internet Movie Data Base and has a number of facts about nearly every movie ever made. It also makes all of its data freely available. (See http://www.imdb.com/interfaces). The data on the site is in tons of different files and is quite a mess, but there is a nice set of scripts from https://github.com/hadley/datamovies. However, I couldn’t quite get that to work, so used an established dataset created using this method and is attached to this quiz called `movies.csv`.

**( a ) Load the file in and double check that all of the columns have the correct type. If they don’t reread in
the file and force the type.**

In [1]:
using DataFrames

In [2]:
movies = readtable("movies.csv",separator=',');

In [3]:
movies

58788×24 DataFrames.DataFrame
│ Row   │ title                      │ year │ length │ budget   │ rating │
├───────┼────────────────────────────┼──────┼────────┼──────────┼────────┤
│ 1     │ "\$"                        │ 1971 │ 121    │ NA       │ 6.4    │
│ 2     │ "\$1000 a Touchdown"        │ 1939 │ 71     │ NA       │ 6.0    │
│ 3     │ "\$21 a Day Once a Month"   │ 1941 │ 7      │ NA       │ 8.2    │
│ 4     │ "\$40,000 "                 │ 1996 │ 70     │ NA       │ 8.2    │
│ 5     │ "\$50,000 Climax Show, The" │ 1975 │ 71     │ NA       │ 3.4    │
│ 6     │ "\$pent"                    │ 2000 │ 91     │ NA       │ 4.3    │
│ 7     │ "\$windle"                  │ 2002 │ 93     │ NA       │ 5.3    │
│ 8     │ "'15'"                     │ 2002 │ 25     │ NA       │ 6.7    │
│ 9     │ "'38"                      │ 1987 │ 97     │ NA       │ 6.6    │
│ 10    │ "'49-'17"                  │ 1917 │ 61     │ NA       │ 6.0    │
│ 11    │ "'68"                      │ 1988 │ 99     │ NA       │ 5.4    │
⋮
│ 58777 │ "eRATicate"                │ 2003 │ 9      │ NA       │ 6.0    │
│ 58778 │ "eXXXorcismos"             │ 2002 │ 78     │ NA       │ 4.2    │
│ 58779 │ "eXistenZ"                 │ 1999 │ 97     │ NA       │ 6.7    │
│ 58780 │ "f2point8"                 │ 2002 │ 20     │ NA       │ 5.0    │
│ 58781 │ "f8"                       │ 2001 │ 13     │ NA       │ 7.6    │
│ 58782 │ "pURe kILLjoy"             │ 1998 │ 87     │ NA       │ 5.2    │
│ 58783 │ "sIDney"                   │ 2002 │ 15     │ NA       │ 7.0    │
│ 58784 │ "tom thumb"                │ 1958 │ 98     │ NA       │ 6.5    │
│ 58785 │ "www.XXX.com"              │ 2003 │ 105    │ NA       │ 1.1    │
│ 58786 │ "www.hellssoapopera.com"   │ 1999 │ 100    │ NA       │ 6.6    │
│ 58787 │ "xXx"                      │ 2002 │ 132    │ 85000000 │ 5.5    │
│ 58788 │ "xXx: State of the Union"  │ 2005 │ 101    │ 87000000 │ 3.9    │

│ Row   │ votes │ r1   │ r2   │ r3   │ r4   │ r5   │ r6   │ r7   │ r8   │ r9   │
├───────┼───────┼──────┼──────┼──────┼──────┼──────┼──────┼──────┼──────┼──────┤
│ 1     │ 348   │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 24.5 │ 24.5 │ 14.5 │ 4.5  │
│ 2     │ 20    │ 0.0  │ 14.5 │ 4.5  │ 24.5 │ 14.5 │ 14.5 │ 14.5 │ 4.5  │ 4.5  │
│ 3     │ 5     │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 24.5 │ 0.0  │ 44.5 │ 24.5 │
│ 4     │ 6     │ 14.5 │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 34.5 │
│ 5     │ 17    │ 24.5 │ 4.5  │ 0.0  │ 14.5 │ 14.5 │ 4.5  │ 0.0  │ 0.0  │ 0.0  │
│ 6     │ 45    │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 14.5 │ 14.5 │ 4.5  │ 4.5  │ 14.5 │
│ 7     │ 200   │ 4.5  │ 0.0  │ 4.5  │ 4.5  │ 24.5 │ 24.5 │ 14.5 │ 4.5  │ 4.5  │
│ 8     │ 24    │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 14.5 │ 14.5 │ 4.5  │
│ 9     │ 18    │ 4.5  │ 4.5  │ 4.5  │ 0.0  │ 0.0  │ 0.0  │ 34.5 │ 14.5 │ 4.5  │
│ 10    │ 51    │ 4.5  │ 0.0  │ 4.5  │ 4.5  │ 4.5  │ 44.5 │ 14.5 │ 4.5  │ 4.5  │
│ 11    │ 23    │ 4.5  │ 0.0  │ 4.5  │ 14.5 │ 24.5 │ 4.5  │ 24.5 │ 4.5  │ 14.5 │
⋮
│ 58777 │ 5     │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 44.5 │ 0.0  │ 0.0  │ 24.5 │
│ 58778 │ 11    │ 34.5 │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 4.5  │ 0.0  │ 14.5 │ 4.5  │
│ 58779 │ 14742 │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 24.5 │ 24.5 │ 14.5 │
│ 58780 │ 13    │ 4.5  │ 0.0  │ 4.5  │ 4.5  │ 14.5 │ 0.0  │ 14.5 │ 4.5  │ 14.5 │
│ 58781 │ 7     │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 24.5 │ 14.5 │ 24.5 │ 24.5 │
│ 58782 │ 6     │ 0.0  │ 14.5 │ 14.5 │ 14.5 │ 0.0  │ 34.5 │ 0.0  │ 0.0  │ 0.0  │
│ 58783 │ 8     │ 14.5 │ 0.0  │ 0.0  │ 14.5 │ 0.0  │ 0.0  │ 24.5 │ 14.5 │ 14.5 │
│ 58784 │ 274   │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 14.5 │ 24.5 │ 14.5 │ 4.5  │
│ 58785 │ 12    │ 45.5 │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 24.5 │ 0.0  │ 0.0  │
│ 58786 │ 5     │ 24.5 │ 0.0  │ 24.5 │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 0.0  │ 24.5 │
│ 58787 │ 18514 │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 14.5 │ 14.5 │ 14.5 │ 4.5  │
│ 58788 │ 1584  │ 24.5 │ 4.5  │ 4.5  │ 4.5  │ 4.5  │ 14.5 │ 4.5  │ 4.5  │ 4.5  │

│ Row   │ r10  │ mpaa    │ Action │ Animation │ Comedy │ 

**( b ) Find the top rated movie(s) and year(s) that it came out. If there is a tie for first, list all movies.**

In [4]:
top_rated = sort!(movies,cols=[:rating],rev=true)[1:30,:]

30×24 DataFrames.DataFrame
│ Row │ title                                     │ year │ length │ budget │
├─────┼───────────────────────────────────────────┼──────┼────────┼────────┤
│ 1   │ "Dimensia Minds Trilogy: The Hope Factor" │ 2004 │ 10     │ NA     │
│ 2   │ "Fishing for Love"                        │ 2001 │ 7      │ NA     │
│ 3   │ "Summer Sonata, A"                        │ 2004 │ 30     │ 2000   │
│ 4   │ "Black Canyon"                            │ 2004 │ 25     │ 3000   │
│ 5   │ "Buck Privates"                           │ 1928 │ 70     │ NA     │
│ 6   │ "Defilada"                                │ 1989 │ 60     │ NA     │
│ 7   │ "Dimensia Minds Trilogy: The Reds"        │ 2004 │ 8      │ 1200   │
│ 8   │ "Drifting"                                │ 2004 │ 21     │ 3000   │
│ 9   │ "Ekdromi, I"                              │ 1966 │ 85     │ NA     │
│ 10  │ "Genet parle d'Angela Davis"              │ 1970 │ 10     │ NA     │
│ 11  │ "Himala"                                  │ 1982 │ 124    │ NA     │
⋮
│ 19  │ "Shades"                                  │ 2004 │ 4      │ NA     │
│ 20  │ "Shopping Part, The"                      │ 2004 │ 12     │ NA     │
│ 21  │ "Sky of an Archangel"                     │ 2004 │ 24     │ NA     │
│ 22  │ "Titanic vals"                            │ 1964 │ 92     │ NA     │
│ 23  │ "Tosca"                                   │ 1956 │ 112    │ NA     │
│ 24  │ "Cashback"                                │ 2004 │ 18     │ NA     │
│ 25  │ "Ceniciento, El"                          │ 1952 │ 95     │ NA     │
│ 26  │ "Challenge"                               │ 1984 │ 152    │ NA     │
│ 27  │ "Chibideka monogatari"                    │ 1958 │ 51     │ NA     │
│ 28  │ "Clase z tropical"                        │ 2000 │ 6      │ NA     │
│ 29  │ "Click Three Times"                       │ 1999 │ 32     │ NA     │
│ 30  │ "Clown"                                   │ 1969 │ 11     │ NA     │

│ Row │ rating │ votes │ r1  │ r2  │ r3  │ r4  │ r5  │ r6  │ r7   │ r8  │ r9   │
├─────┼────────┼───────┼─────┼─────┼─────┼─────┼─────┼─────┼──────┼─────┼──────┤
│ 1   │ 10.0   │ 5     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 0.0  │
│ 2   │ 10.0   │ 5     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 0.0  │
│ 3   │ 10.0   │ 5     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 0.0  │
│ 4   │ 9.9    │ 9     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 5   │ 9.9    │ 7     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 6   │ 9.9    │ 8     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 7   │ 9.9    │ 15    │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 8   │ 9.9    │ 41    │ 0.0 │ 4.5 │ 4.5 │ 0.0 │ 0.0 │ 4.5 │ 4.5  │ 0.0 │ 0.0  │
│ 9   │ 9.9    │ 7     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 10  │ 9.9    │ 7     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 24.5 │
│ 11  │ 9.9    │ 26    │ 4.5 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 4.5  │
⋮
│ 19  │ 9.9    │ 11    │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 4.5  │
│ 20  │ 9.9    │ 7     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 21  │ 9.9    │ 7     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │
│ 22  │ 9.9    │ 13    │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 4.5  │ 4.5 │ 4.5  │
│ 23  │ 9.9    │ 10    │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 14.5 │ 0.0 │ 24.5 │
│ 24  │ 9.8    │ 51    │ 4.5 │ 0.0 │ 0.0 │ 4.5 │ 4.5 │ 0.0 │ 0.0  │ 4.5 │ 4.5  │
│ 25  │ 9.8    │ 8     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 24.5 │
│ 26  │ 9.8    │ 8     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 24.5 │
│ 27  │ 9.8    │ 8     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 14.5 │ 0.0 │ 14.5 │
│ 28  │ 9.8    │ 5     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 24.5 │
│ 29  │ 9.8    │ 5     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 24.5 │
│ 30  │ 9.8    │ 6     │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0 │ 0.0  │ 0.0 │ 14.5 │

│ Row │ r10   │ m

**( c ) Create a `Dataframe` with only movies with more than 10000 votes. Find the top 5 rated movies(s) and year(s) of these movie.**

In [5]:
high_votes = by(movies[10000 .< movies[:votes],:],:title) do df # df is dataframe
    DataFrame(Year=df[:year],Votes=sum(df[:votes]))
end
high_votes = sort!(high_votes,cols=[:Votes],rev=true)[1:5,:]

5×3 DataFrames.DataFrame
│ Row │ title                                                │ Year │ Votes  │
├─────┼──────────────────────────────────────────────────────┼──────┼────────┤
│ 1   │ "Lord of the Rings: The Fellowship of the Ring, The" │ 2001 │ 157608 │
│ 2   │ "Shawshank Redemption, The"                          │ 1994 │ 149494 │
│ 3   │ "Matrix, The"                                        │ 1999 │ 143853 │
│ 4   │ "Star Wars"                                          │ 1977 │ 134640 │
│ 5   │ "Pulp Fiction"                                       │ 1994 │ 132745 │

**( d ) Create another dataframe that consists of only movies from the 1980s. Find the number of movies with a rating of 9.0 or better.**

In [6]:
movies_80s = by(movies[1980 .<= movies[:year] .< 1990,:],:title) do df # df is dataframe
    DataFrame(Year=df[:year],Rating=df[:rating])
end
k=0
for i=1:size(movies_80s)[1]
    movies_80s[i,:Rating] >= 9.0 ? k+=1 : k=k
end
k

125

There are 125 movies with rating 9.0 or better

**( e ) What was the longest movie (in running time) in the 1980s.**

In [7]:
movies_80s_by_length = by(movies[1980 .<= movies[:year] .< 1990,:],:title) do df # df is dataframe
    DataFrame(Year=df[:year],Runtime=df[:length])
end
longest_movie = sort!(movies_80s_by_length,cols=[:Runtime],rev=true)[1,:]

1×3 DataFrames.DataFrame
│ Row │ title                    │ Year │ Runtime │
├─────┼──────────────────────────┼──────┼─────────┤
│ 1   │ "Cure for Insomnia, The" │ 1987 │ 5220    │

In [8]:
5220/60/24

3.625

In [9]:
.625*24

15.0

*The Cure for Insomnia* which ran 5220 minutes or 3 days 15 hours

**( f ) Split the movies of the 1980s into individual years. Combine the result by average rating. In the end, you should have a table of size 10 with only the year and average rating.**

In [10]:
movies_80s_by_rating_and_year = by(movies_80s,:Year) do df
    DataFrame(Average_Rating=sum(df[:Rating])/(length(df[:Year])))
end

10×2 DataFrames.DataFrame
│ Row │ Year │ Average_Rating │
├─────┼──────┼────────────────┤
│ 1   │ 1980 │ 5.62482        │
│ 2   │ 1981 │ 5.68064        │
│ 3   │ 1982 │ 5.77329        │
│ 4   │ 1983 │ 5.66719        │
│ 5   │ 1984 │ 5.73458        │
│ 6   │ 1985 │ 5.66452        │
│ 7   │ 1986 │ 5.57197        │
│ 8   │ 1987 │ 5.55235        │
│ 9   │ 1988 │ 5.50042        │
│ 10  │ 1989 │ 5.54047        │

## Problem 2:
Adapt your Geometry module from Homeworks #5 and #6. (you can make a copy of that module).

**( a ) Add another type called `Triangle` that has three points.**
> Added to module

**( b ) Write a function called `is_equilateral` that takes a `Triangle` as an input and determines if the input is an equilateral triangle.**

In [11]:
if !("." in LOAD_PATH)
    push!(LOAD_PATH,".");
end

3-element Array{String,1}:
 "/Applications/Julia-0.5.app/Contents/Resources/julia/local/share/julia/site/v0.5"
 "/Applications/Julia-0.5.app/Contents/Resources/julia/share/julia/site/v0.5"      
 "."                                                                               

In [12]:
using Geometry

In [13]:
t1 = Triangle(Point(0,0),Point(4,0),Point(4,3))

Geometry.Triangle(Geometry.Point(0,0),Geometry.Point(4,0),Geometry.Point(4,3))

In [14]:
t2 = Triangle(Point(-2,0),Point(2,0),Point(0,2*sqrt(3)))

Geometry.Triangle(Geometry.Point(-2,0),Geometry.Point(2,0),Geometry.Point(0,3.4641016151377544))

In [15]:
is_equilateral(t1)

false

In [16]:
is_equilateral(t2)

true

**( c ) Write a function called `is_right` that takes a `Triangle` as an input and determines if the input is a right triangle. (Hint: recall Pythagoras’ Theorem.)**

In [17]:
is_right(t1)

true

In [18]:
is_right(t2)

false

**( d ) There is a file called `test-triangle.jl`, run it as is to test these functions.**

In [19]:
include("test-triangle.jl")

Test Summary:  | Pass  Total
  Triangle Tests |   12     12


Base.Test.DefaultTestSet("Triangle Tests",Any[Base.Test.DefaultTestSet("Triangle Constructor",Any[Test Passed
  Expression: isa(t1,Triangle),Test Passed
  Expression: isa(t2,Triangle),Test Passed
  Expression: isa(t3,Triangle),Test Passed
  Expression: isa(t4,Triangle)],false),Base.Test.DefaultTestSet("Equilateral Triangles",Any[Test Passed
  Expression: !(is_equilateral(t1)),Test Passed
  Expression: is_equilateral(t2),Test Passed
  Expression: !(is_equilateral(t3)),Test Passed
  Expression: !(is_equilateral(t4))],false),Base.Test.DefaultTestSet("Right Triangle Tests",Any[Test Passed
  Expression: is_right(t1),Test Passed
  Expression: !(is_right(t2)),Test Passed
  Expression: is_right(t3),Test Passed
  Expression: !(is_right(t4))],false)],false)